build/build-image/cross/Makefile

```diff
@@ -12,16 +12,85 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-.PHONY:	build push
+.PHONY: all push container clean orphan all-push push-manifest
+
+include ../../../hack/make-rules/Makefile.manifest
+
+REGISTRY ?= staging-k8s.gcr.io
+IMAGE = $(REGISTRY)/kube-cross
+IMAGE_WITH_ARCH = $(IMAGE)-$(ARCH)
+
+# This option is for running docker manifest command
+export DOCKER_CLI_EXPERIMENTAL := enabled
 
-IMAGE=kube-cross
 TAG=$(shell cat VERSION)
 
+# Architectures supported: amd64, arm64, ppc64le
+ARCH ?= amd64
+
+ALL_ARCH = amd64 arm64 ppc64le
+
+TEMP_DIR:=$(shell mktemp -d)
+QEMUVERSION=v2.9.1
+
+ifeq ($(ARCH),arm64)
+	QEMUARCH=aarch64
+endif
+ifeq ($(ARCH),ppc64le)
+	QEMUARCH=ppc64le
+endif
+
+SUDO=$(if $(filter 0,$(shell id -u)),,sudo)
+
+# If you want to build AND push all containers, see the 'all-push' rule.
+all: all-container
+
+all-push: all-push-images push-manifest
+
+push-manifest:
+	docker manifest create --amend $(IMAGE):$(TAG) $(shell echo $(ALL_ARCH) | sed -e "s~[^ ]*~$(IMAGE)\-&:$(TAG)~g")
+	@for arch in $(ALL_ARCH); do docker manifest annotate --arch $${arch} ${IMAGE}:${TAG} ${IMAGE}-$${arch}:${TAG}; done
+	docker manifest push --purge ${IMAGE}:${TAG}
+
+sub-container-%:
+	$(MAKE) ARCH=$* container
+
+sub-push-%:
+	$(MAKE) ARCH=$* push
+
+all-container: $(addprefix sub-container-,$(ALL_ARCH))
+
+all-push-images: $(addprefix sub-push-,$(ALL_ARCH))
+
+container: .container-$(ARCH)
+.container-$(ARCH):
+	cp ./* $(TEMP_DIR)
+	cat Dockerfile \
+		| sed "s|QEMUARCH|$(QEMUARCH)|g" \
+		> $(TEMP_DIR)/Dockerfile
+
+ifeq ($(ARCH),amd64)
+	# When building "normally" for amd64, remove the whole line, it has no part in the amd64 image
+	sed "/CROSS_BUILD_/d" $(TEMP_DIR)/Dockerfile > $(TEMP_DIR)/Dockerfile.tmp
+else
+	# When cross-building, only the placeholder "CROSS_BUILD_" should be removed
+	# Register /usr/bin/qemu-ARCH-static as the handler for non-x86 binaries in the kernel
+	$(SUDO) ../../../third_party/multiarch/qemu-user-static/register/register.sh --reset
```

Comment:
I don't like this. Building cross now requires root?

Reply:
we use this for the same thing for all the images ... :(

Reply:
cross compiling has always required this IIRC, just root in a container usually.

Reply:
This breaks the build under Darwin:
```
# When cross-building, only the placeholder "CROSS_BUILD_" should be removed
# Register /usr/bin/qemu-ARCH-static as the handler for non-x86 binaries in the kernel
sudo ../../../third_party/multiarch/qemu-user-static/register/register.sh --reset
No binfmt support in the kernel.
  Try: '/sbin/modprobe binfmt_misc' from the host
make[1]: *** [.container-arm64] Error 1
make: *** [sub-container-arm64] Error 2
```